x-pack/plugins/task_manager/server/integration_tests/task_manager_capacity_based_claiming.test.ts

```diff
@@ -0,0 +1,343 @@
+          capacity: 10,
+          poll_interval: POLLING_INTERVAL,
+          unsafe: {
+            exclude_task_types: [
+              'security*',
+              'osquery*',
+              'endpoint*',
+              'apm*',
+              'ML*',
+              'task_manager*',
+              'alert*',
+              'action*',
+              'cases*',
+              'SLO*',
+              'Fleet*',
+              'fleet*',
+              'session*',
+              'dashboard*',
+              'observability*',
+            ],
+          },
+        },
+      },
+    });
+    esServer = setupResult.esServer;
+    kibanaServer = setupResult.kibanaServer;
+
+    expect(taskManagerStartSpy).toHaveBeenCalledTimes(1);
+    taskManagerPlugin = taskManagerStartSpy.mock.results[0].value;
+
+    expect(TaskPollingLifecycleMock).toHaveBeenCalledTimes(1);
+
+    expect(createMonitoringStatsMock).toHaveBeenCalledTimes(1);
+    createMonitoringStatsOpts = createMonitoringStatsMock.mock.calls[0][0];
+  });
+
+  afterAll(async () => {
+    if (kibanaServer) {
+      await kibanaServer.stop();
+    }
+    if (esServer) {
+      await esServer.stop();
+    }
+  });
+
+  beforeEach(() => {
+    jest.clearAllMocks();
+  });
+
+  afterEach(async () => {
+    while (taskIdsToRemove.length > 0) {
+      const id = taskIdsToRemove.pop();
+      await taskManagerPlugin.removeIfExists(id!);
+    }
+  });
+
+  it('should claim tasks to full capacity', async () => {
+    const backgroundTaskLoads: number[] = [];
+    createMonitoringStatsOpts.taskPollingLifecycle?.events
+      .pipe(
+        filter(isTaskManagerWorkerUtilizationStatEvent),
+        map<TaskLifecycleEvent, number>((taskEvent: TaskLifecycleEvent) => {
+          return (taskEvent.event as unknown as Ok<number>).value;
+        })
+      )
+      .subscribe((load: number) => {
+        backgroundTaskLoads.push(load);
+      });
+    const taskRunAtDates: Date[] = [];
+    mockTaskTypeNormalCostRunFn.mockImplementation(() => {
+      taskRunAtDates.push(new Date());
+      return { state: { foo: 'test' } };
+    });
+
+    // inject 10 normal cost tasks with the same runAt value
+    const ids: string[] = [];
+    times(10, () => ids.push(uuidV4()));
+
+    const runAt = new Date();
+    for (const id of ids) {
+      await injectTask(kibanaServer.coreStart.elasticsearch.client.asInternalUser, {
+        id,
+        taskType: 'normalCostType',
+        params: {},
+        state: { foo: 'test' },
+        stateVersion: 1,
+        runAt,
+        enabled: true,
+        scheduledAt: new Date(),
+        attempts: 0,
+        status: TaskStatus.Idle,
+        startedAt: null,
+        retryAt: null,
+        ownerId: null,
+      });
+      taskIdsToRemove.push(id);
+    }
+
+    await retry(async () => {
+      expect(mockTaskTypeNormalCostRunFn).toHaveBeenCalledTimes(10);
+    });
+
+    expect(taskRunAtDates.length).toBe(10);
+
+    // run at dates should be within a few seconds of each other
```

Comment:
should be within milliseconds, but I'm adding some fudge factor

Reply:
We shall see how much fudge we will need to apply! :-)